src/number_theory/legendre_symbol/quadratic_char.lean

```diff
@@ -0,0 +1,339 @@
+/-
+Copyright (c) 2022 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Stoll
+-/
+import tactic.basic
+import field_theory.finite.basic
+
+/-!
+# Quadratic characters of finite fields
+
+This file defines the quadratic character on a finite field `F` and proves
+some basic statements about it.
+
+## Tags
+
+quadratic character
+-/
+
+/-!
+### Some general results on finite fields
+-/
+
+section general
+
+/-- Create a decidability instance for `is_square` on finite monoids. -/
+instance {M : Type*} [fintype M] [monoid M] [decidable_eq M] :
+  decidable_pred (is_square : M → Prop) :=
+λ a, decidable_of_iff' _ (is_square_iff_exists_sq a)
+
+/-- `0` is always a square (in a monoid with zero). -/
+lemma is_square_zero (M : Type*) [monoid_with_zero M] : is_square (0 : M) :=
+by { use 0, simp only [mul_zero] }
+
+/-- If `ring_char R = 2`, where `R` is a finite reduced commutative ring,
+then every `a : R` is a square. -/
+lemma is_square_of_char_two' {R : Type*} [fintype R] [comm_ring R] [is_reduced R] [char_p R 2]
+ (a : R) : is_square a :=
+exists_imp_exists (λ b h, pow_two b ▸ eq.symm h) $
+  ((fintype.bijective_iff_injective_and_card _).mpr ⟨frobenius_inj R 2, rfl⟩).surjective a
```

Comment:
Could you please move these to near the definition of `is_square`? I guess all the imports you need will already be available there?

Reply:
The decidability instance will be more general (requiring only `has_mul` instead of `monoid`). See the [discussion starting here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Quadratic.20Hilbert.20symbol.20over.20.E2.84.9A/near/279714288).

Reply:
Ok, it's fine to leave the last one here for now.